### ISSUE TITLE
time options

### ISSUE THREAD

**Issue by ltdeta** [Arkni/date-validator#2]
is it possible extend dateMultipleFormat with `time options`
for a compatible with [jQuery-Timepicker-Addon](https://github.com/trentrichardson/jQuery-Timepicker-Addon)

**Comment by Arkni** [comment]
I think you mean, adding support to validate a date-time value, right ?

 or you want to validate either the date or the time ? if yes, there is already a validator for time in the [additional methods](https://github.com/jzaefferer/jquery-validation/blob/master/src/additional/):
- time: https://github.com/jzaefferer/jquery-validation/blob/master/src/additional/time.js
- time12h: https://github.com/jzaefferer/jquery-validation/blob/master/src/additional/time12h.js

**Comment by ltdeta** [comment]
no i want to validate date and time together
have a look at [timepicker](http://trentrichardson.com/examples/timepicker/#tp-formatting)

example: 01.12.2012 12:10

**Comment by Arkni** [comment]
Please check out this [branch](https://github.com/Arkni/date-validator/tree/time-support) and test it.

If there any problem, please report.

**Comment by ltdeta** [comment]
my experience
- `function parse` not acepting time data
- line 247 comma is wrong should be a semicolon

``` code
formats     = format.split( " " ),
```
- `function timeHelper` not validating if minutes or second

request: 
support time formating
- Hour with leading 0  -  2:12  =>  02:12
- Minutes with leading 0  -  12:1  =>  12:01
- seconds with leading 0  -  12:03:4  =>  12:03:04

**Comment by Arkni** [comment]
> `function parse` not acepting time data.

Yeh, I missed that point. I will fix that soon.

> `function timeHelper` not validating if minutes or second

The time format should consist of three parts:
- Hours
- Minutes
- Seconds

**Comment by ltdeta** [comment]
> The time format should consist of three parts 

**example `format` : "DD.MM.YYYY HH:mm:ss"**
01.10.2014 12:   => should invalid 
01.10.2014 12:22   => should invalid
01.10.2014 12:22:   => should invalid
01.10.2014 12:01:12   => valid

**example `format` : "DD.MM.YYYY HH:mm"**
01.10.2014 12:   => should invalid 
01.10.2014 12:22:   => should invalid
01.10.2014 12:01:12   => should invalid
01.10.2014 12:22   =>  valid

**Comment by Arkni** [comment]
> **example `format` : "DD.MM.YYYY HH:mm:ss"**
> 01.10.2014 12:   => should invalid 
> 01.10.2014 12:22   => should invalid
> 01.10.2014 12:22:   => should invalid
> 01.10.2014 12:01:12   => valid
> 
> **example `format` : "DD.MM.YYYY HH:mm"**
> 01.10.2014 12:   => should invalid 
> 01.10.2014 12:22:   => should invalid
> 01.10.2014 12:01:12   => should invalid
> 01.10.2014 12:22   =>  valid

Please check the new changes: 9dc294f3f5be4a4001b39c82682f397049345bf6

**Comment by Arkni** [comment]
I'm working now on accepting all these tokens:

| Token | Description |
| --- | --- |
| D | Day of the month as digits; no leading zero for single-digit days. |
| DD | Day of the month as digits; leading zero for single-digit days. |
| M | Month as digits; no leading zero for single-digit months. |
| MM | Month as digits; leading zero for single-digit months. |
| YY | Year as last two digits; leading zero for years less than 10. |
| YYYY | Year represented by four digits between 1000 and 9999. |
| h | Hours; no leading zero for single-digit hours (12-hour clock AM or PM). |
| hh | Hours; leading zero for single-digit hours (12-hour clock AM or PM). |
| H | Hours; no leading zero for single-digit hours (24-hour clock). |
| HH | Hours; leading zero for single-digit hours (24-hour clock). |
| mm | Minutes; no leading zero for single-digit minutes. |
| mm | Minutes; leading zero for single-digit minutes. |
| s | Seconds; no leading zero for single-digit seconds. |
| ss | Seconds; leading zero for single-digit seconds. |
| A | AM PM |
| a | am pm |

**# Notes:**
- `A` and `a` should be used with `hh` and `h`

**Comment by ltdeta** [comment]
nice

**Comment by Arkni** [comment]
@ltdeta 
If you have the time, Could you please test the new changes: e2e4108b3ff56b41f8e4c645a25d0a9484481f2f

**Comment by ltdeta** [comment]
all works fine thanks

 i modified the error message handling  for my purposes
inserting mutiple errormessages for errors like format, mindate, maxdate and min-max not in range

``` javascript
$.extend($.validator.messages, {
.......
    dateMultipleFormatMinMax: $.validator.format("Das Datum muss im Zeitraum {0} - {1} liegen."),
    dateMultipleFormatMin: $.validator.format("Das Datum muss größer gleich {0} sein."),
    dateMultipleFormatMax: $.validator.format("Das Datum muss kleiner gleich {0} sein."),
    dateMultipleFormatFormat: $.validator.format("Das Datum bitte im Format {0} eingegeben.")
});
```

``` javascript
$.validator.addMethod('xxxxx', function (val, element) {
    // do stuff
   // extend the options with a property result 
}} , function(options, element) {//, $.validator.messages.date);
          var retMsg = "";
          switch ( options.result ) {
                      case 2:
                  retMsg = $.validator.messages.dateMultipleFormatMin([options.minDate]);
                  break;
              case 3:
                  retMsg = $.validator.messages.dateMultipleFormatMax([options.maxDate]);
                  break;
              case 4:
                retMsg = $.validator.messages.dateMultipleFormatMinMax([options.minDate,options.maxDate]);
                break;
              default:
                retMsg = $.validator.messages.dateMultipleFormatFormat([options.format]);
          }
          return retMsg;
    });
});
```

**Comment by Arkni** [comment]
Thanks a lot for testing it.

> i modified the error message handling for my purposes

I was wondering how to do something like that, thanks for sharing :)

**Comment by Arkni** [comment]
@ltdeta 
Can you point me please to an article or a piece of documentation that describe the method (the way to use multiple messages) you're using ?

**Comment by ltdeta** [comment]
[Dynamic jQuery Validate error messages with AddMethod based on the element](http://stackoverflow.com/questions/13352626/dynamic-jquery-validate-error-messages-with-addmethod-based-on-the-element)

[jquery validator addmethod custom message](http://gotoanswer.com/?q=jquery+validator+addmethod+custom+message)

**Comment by Arkni** [comment]
Thanks a lot :)
I appreciate your help :)

**Comment by Arkni** [comment]
I ended up using this way instead:

``` diff
diff --git a/src/date-validator.js b/src/date-validator.js
index 71e3829..6252fe2 100644
--- a/src/date-validator.js
+++ b/src/date-validator.js
@@ -322,6 +322,67 @@
        return true;
    }

+   /**
+    * This function is based on the dateFormat function from the Date Format 1.2.3
+    * Credit to (c) 2007-2009 Steven Levithan <stevenlevithan.com>
+    * MIT license
+    * see http://blog.stevenlevithan.com/archives/date-time-format for the complete lib
+    *
+    * Return the date string formatted following the format provided as param
+    *
+    * @param {Date} date      The date object to format
+    * @param {String} format  The date format
+    *
+    * @returns {String}
+    */
+   function dateFormater( date, format ) {
+       var token = /D{1,2}|M{1,2}|YY(?:YY)?|([HhmsAa])\1?|"[^"]*"|'[^']*'/g,
+           pad = function( val ) {
+               val = "" + val;
+               while ( val.length < 2 ) {
+                   val = "0" + val;
+               }
+               return val;
+           },
+           amOrPm = function( hours, token) {
+               if ( token === "a" ) {
+                   return hours <= 12 ? "am" : "pm";
+               }
+               return hours <= 12 ? "AM" : "PM";
+           }, flags;
+
+       // Passing date through Date applies Date.parse, if necessary
+       date = date ? new Date(date) : new Date();
+
+       // This is not supposed to happen, but play it safe.
+       if ( isNaN( date ) ) {
+           throw new SyntaxError( "invalid date" );
+       }
+
+       flags = {
+           D:    date.getDate(),
+           DD:   pad( date.getDate() ),
+           M:    date.getMonth() + 1,
+           MM:   pad( date.getMonth() + 1 ),
+           YY:   String( date.getFullYear() ).slice( 2 ),
+           YYYY: date.getFullYear(),
+           h:    date.getHours() % 12 || 12,
+           hh:   pad( date.getHours() % 12 || 12 ),
+           H:    date.getHours(),
+           HH:   pad( date.getHours() ),
+           m:    date.getMinutes(),
+           mm:   pad( date.getMinutes() ),
+           s:    date.getSeconds(),
+           ss:   pad( date.getSeconds() ),
+           a:    amOrPm( date.getHours(), "a" ),
+           A:    amOrPm( date.getHours(), "A" )
+       };
+
+       return format.replace( token, function( match ) {
+           return match in flags ? flags[ match ] : match.slice( 1, match.length - 1 );
+       });
+   }
+
    /*
     * dateValidator validator:
     *     check if the given date is valid following the given format.
@@ -337,6 +398,9 @@
     * Usage examples & docs: http://arkni.github.io/date-validator/
     */
    $.validator.addMethod("dateValidator", function( value, element, options ) {
+       // Set the default message, the one used by the build in date validator.
+       $.validator.messages.dateValidator = $.validator.messages.date;
+
        // If the field is optional
        // exit with success
        if ( this.optional( element ) ) {
@@ -416,12 +480,28 @@
            switch ( true ) {
            case ( options.minDate && !options.maxDate && valid ):
                valid = date.getTime() >= minDate.getTime();
+
+               // Format the minDate if it is an instance of Date object
+               minDate = minDate instanceof Date ? dateFormater( minDate, format ) : minDate;
+               // Set the new error message based on the minDate value.
+               $.validator.messages.dateValidator = $.validator.format( "Please enter a date after {0}.", minDate );
                break;
            case ( options.maxDate && !options.minDate && valid ):
                valid = date.getTime() <= maxDate.getTime();
+
+               // Format the maxDate if it is an instance of Date object
+               maxDate = maxDate instanceof Date ? dateFormater( maxDate, format ) : maxDate;
+               // Set the new error message based on the maxDate value.
+               $.validator.messages.dateValidator = $.validator.format( "Please enter a date before {0}.", maxDate );
                break;
            case ( options.minDate && options.maxDate && valid ):
                valid = date.getTime() >= minDate.getTime() && date.getTime() <= maxDate.getTime();
+
+               // Format the minDate & the maxDate if they are instances of Date object
+               minDate = minDate instanceof Date ? dateFormater( minDate, format ) : minDate;
+               maxDate = maxDate instanceof Date ? dateFormater( maxDate, format ) : maxDate;
+               // Set the new error message based on the minDate & maxDate values.
+               $.validator.messages.dateValidator = $.validator.format( "Please enter a date between {0} and {1}.", minDate, maxDate );
                break;
            default:
                break;
@@ -429,5 +509,5 @@
        }

        return valid;
-   }, $.validator.messages.date);
+   }, $.validator.messages.dateValidator);
 }());
```

I'll push these chages soon, and then create a new release.

**Comment by ltdeta** [comment]
Ok, but i believe "my way" is more flexible using other languages 
because i must not change the dateValidator itself.

**Comment by Arkni** [comment]
I see your point now. I don't know how I missed that.

What about using something like this (It's based on your code):

``` diff
diff --git a/src/date-validator.js b/src/date-validator.js
index fb6713a..9c4a537 100644
--- a/src/date-validator.js
+++ b/src/date-validator.js
@@ -398,9 +398,6 @@
     * Usage examples & docs: http://arkni.github.io/date-validator/
     */
    $.validator.addMethod("dateValidator", function( value, element, options ) {
-       // Set the default message, the one used by the build in date validator.
-       $.validator.messages.dateValidator = $.validator.messages.date;
-
        // If the field is optional
        // exit with success
        if ( this.optional( element ) ) {
@@ -450,6 +447,9 @@
            timeValue  = ( values.length > 1 ) ? values[ 1 ] : null;
            amOrPmValue = ( values.length > 2 ) ? values[ 2 ] : null;

+           // Set the default message, the one used by the build in date validator.
+           $.validator.messages.dateValidator = $.validator.messages.dateValidatorMessages[ "default" ]( format );
+
            // Check if formats & values have the same length
            // If no, exit with error
            if ( formats.length !== values.length ) {
@@ -484,7 +484,7 @@
                // Format the minDate if it is an instance of Date object
                minDate = minDate instanceof Date ? dateFormater( minDate, format ) : minDate;
                // Set the new error message based on the minDate value.
-               $.validator.messages.dateValidator = $.validator.format( "Please enter a date after {0}.", minDate );
+               $.validator.messages.dateValidator = $.validator.messages.dateValidatorMessages.minDate( minDate );
                break;
            case ( options.maxDate && !options.minDate && valid ):
                valid = date.getTime() <= maxDate.getTime();
@@ -492,7 +492,7 @@
                // Format the maxDate if it is an instance of Date object
                maxDate = maxDate instanceof Date ? dateFormater( maxDate, format ) : maxDate;
                // Set the new error message based on the maxDate value.
-               $.validator.messages.dateValidator = $.validator.format( "Please enter a date before {0}.", maxDate );
+               $.validator.messages.dateValidator = $.validator.messages.dateValidatorMessages.maxDate( maxDate );
                break;
            case ( options.minDate && options.maxDate && valid ):
                valid = date.getTime() >= minDate.getTime() && date.getTime() <= maxDate.getTime();
@@ -501,7 +501,7 @@
                minDate = minDate instanceof Date ? dateFormater( minDate, format ) : minDate;
                maxDate = maxDate instanceof Date ? dateFormater( maxDate, format ) : maxDate;
                // Set the new error message based on the minDate & maxDate values.
-               $.validator.messages.dateValidator = $.validator.format( "Please enter a date between {0} and {1}.", minDate, maxDate );
+               $.validator.messages.dateValidator = $.validator.messages.dateValidatorMessages.range( minDate, maxDate );
                break;
            default:
                break;
@@ -510,4 +510,14 @@

        return valid;
    }, $.validator.messages.dateValidator);
+
+   // Adding the dateValidator messages.
+   $.extend($.validator.messages, {
+       dateValidatorMessages: {
+           "default": $.validator.format( "Please enter a valid date, format {0}."),
+           minDate: $.validator.format( "Please enter a date after {0}."),
+           maxDate: $.validator.format( "Please enter a date before {0}."),
+           range: $.validator.format( "Please enter a date between {0} and {1}.")
+       }
+   });
 }());
```

It doesn't require an extra option (I mean the `result` option that you have added) or the function you've created as well.

**Comment by ltdeta** [comment]
> It doesn't require an extra option
> yes, i also think that

**Comment by ltdeta** [comment]
do you have already published the last changes?

**Comment by Arkni** [comment]
Yes, I did.

**Comment by Arkni** [comment]
all available in master and the new release v1.0.2
